services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java

```diff
@@ -239,16 +273,349 @@ public byte[] encodePassword(byte[] challenge, String password) throws Exception
+            byte[] reasonBuffer = new byte[reasonLength];
+            tunnelInputStream.readFully(reasonBuffer);
+            String reason = new String(reasonBuffer);
+            String errMsg = "No security type provided by the VNC server, reason: " + reason;
+            s_logger.error(errMsg);
+            throw new IOException(errMsg);
+        }
+
+        for (int i = 0; i < numberOfSecurityTypes; i++) {
+            int securityType = tunnelInputStream.readByte();
+            if (securityType != 0 && VncSecurity.supportedSecurityTypes.contains(securityType)) {
+                s_logger.info("Selected the security type: " + securityType);
+                return securityType;
+            }
+        }
+        throw new IOException("Could not select a supported or valid security type from the offered by the server");
+    }
+
+    /**
+     * VNC authentication.
+     */
+    public void processSecurityResult(String password)
+            throws IOException {
+        // Read security result
+        int authResult = this.tunnelInputStream.readInt();
+
+        switch (authResult) {
+            case RfbConstants.VNC_AUTH_OK: {
+                // Nothing to do
+                break;
+            }
+
+            case RfbConstants.VNC_AUTH_TOO_MANY:
+                s_logger.error("Connection to VNC server failed: too many wrong attempts.");
+                throw new RuntimeException("Connection to VNC server failed: too many wrong attempts.");
+
+            case RfbConstants.VNC_AUTH_FAILED:
+                s_logger.error("Connection to VNC server failed: wrong password.");
+                throw new RuntimeException("Connection to VNC server failed: wrong password.");
+
+            default:
+                s_logger.error("Connection to VNC server failed, reason code: " + authResult);
+                throw new RuntimeException("Connection to VNC server failed, reason code: " + authResult);
+        }
+    }
+
     public int read(byte[] b) throws IOException {
-        return is.read(b);
+        return tunnelInputStream.read(b);
     }
 
     public void write(byte[] b) throws IOException {
         if (isVncOverWebSocketConnection()) {
             proxyMsgOverWebSocketConnection(ByteBuffer.wrap(b));
+        } else if (!isVncOverTunnel()) {
+            this.socketConnection.writeBytes(b, 0, b.length);
+        } else {
+            tunnelOutputStream.write(b);
+        }
+    }
+
+    public void writeFrame(Frame frame) {
+        byte[] data = new byte[frame.getPayloadLength()];
+        frame.getPayload().get(data);
+
+        if (securityPhaseCompleted) {
+            socketConnection.writeBytes(ByteBuffer.wrap(data), data.length);
+            socketConnection.flushWriteBuffer();
+            if (writerLeft == null) {
+                writerLeft = 3;
+                setWaitForNoVnc(false);
+            } else if (writerLeft > 0) {
+                writerLeft--;
+            }
```

Comment:
`writeSecureFrame()`?

Reply:
No, insecure or secure use the same nioSocketConnection